### PR TITLE
Develop stream 20220124

### PR DESCRIPTION
**Pull request by mfep**
- Fixed radix merge comparator
- Restored future value tests
   - The tests were ready when the feature was added, but they were lost in a rebase, now restored
- Implemented ranked exchange in warp exchange
- Implemented kernel for device scan-by-key
- Improved small segment sorting in device segmented radix sort
- Implemented test suite for HIP async copy
- Added launch configuration for MI100 for device merge sort and device radix sort

### CODE REVIEW

File: rocprim/include/rocprim/detail/radix_sort.hpp [hunk 1]
```diff
@@ -49,13 +50,17 @@ struct radix_key_codec_integral<Key, BitKey, typename std::enable_if<::rocprim::
     ROCPRIM_DEVICE ROCPRIM_INLINE
     static bit_key_type encode(Key key)
     {
-        return *reinterpret_cast<bit_key_type *>(&key);
+        bit_key_type bit_key;
+        std::memcpy(&bit_key, &key, sizeof(bit_key));
```
[the remark this answers is not in the document]
Comment (mfep) [reply rocprim/include/rocprim/detail/radix_sort.hpp:54]: In that case, I'd suggest to use `__builtin_memcpy` instead, which is already spread throughout the code. I'm pushing the fix in a moment...

[the remark this answers is not in the document]
Comment (stanleytsang-amd) [reply rocprim/include/rocprim/detail/radix_sort.hpp]: Sounds good.  That fixed the build error, however there is another build error in CI that's showing up now, but I am unable to reproduce it on my test machine.  Still looking into it.

**Comment by stanleytsang-amd** [comment]
@mfep  I just tried building rocThrust against this PR and I get compilation errors.  Can you confirm on your end?

In file included from /opt/rocm-5.0.0/include/thrust/sort.h:1358:
In file included from /opt/rocm-5.0.0/include/thrust/detail/sort.inl:26:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/sort.h:152:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/sort.inl:26:
In file included from /opt/rocm-5.0.0/include/thrust/find.h:381:
In file included from /opt/rocm-5.0.0/include/thrust/detail/find.inl:25:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/find.h:61:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/find.inl:19:
In file included from /opt/rocm-5.0.0/include/thrust/reduce.h:781:
In file included from /opt/rocm-5.0.0/include/thrust/detail/reduce.inl:28:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/reduce_by_key.h:87:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/reduce_by_key.inl:37:
In file included from /opt/rocm-5.0.0/include/thrust/scan.h:1594:
In file included from /opt/rocm-5.0.0/include/thrust/detail/scan.inl:28:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/adl/scan.h:44:
In file included from /opt/rocm-5.0.0/include/thrust/system/hip/detail/scan.h:48:
In file included from /opt/rocm-5.0.0/include/rocprim/rocprim.hpp:54:
In file included from /opt/rocm-5.0.0/include/rocprim/device/device_merge.hpp:30:
/opt/rocm-5.0.0/include/rocprim/device/device_merge_config.hpp:47:9: error: no matching function for call to 'ceili                                                                                          ng_div'
        ::rocprim::detail::ceiling_div<unsigned int>(::rocprim::max(sizeof(Key), sizeof(Value)), sizeof(int));
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-5.0.0/include/rocprim/device/device_merge_config.hpp:57:9: error: no matching function for call to 'ceili                                                                                          ng_div'
        ::rocprim::detail::ceiling_div<unsigned int>(sizeof(Key), sizeof(int));
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-5.0.0/include/rocprim/device/device_merge_config.hpp:71:9: error: no matching function for call to 'ceili                                                                                          ng_div'
        ::rocprim::detail::ceiling_div<unsigned int>(::rocprim::max(sizeof(Key), sizeof(Value)), sizeof(int));
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-5.0.0/include/rocprim/device/device_merge_config.hpp:81:9: error: no matching function for call to 'ceili                                                                                          ng_div'
        ::rocprim::detail::ceiling_div<unsigned int>(sizeof(Key), sizeof(int));
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-5.0.0/include/rocprim/device/device_merge_config.hpp:96:9: error: no matching function for call to 'ceili                                                                                          ng_div'
        ::rocprim::detail::ceiling_div<unsigned int>(::rocprim::max(sizeof(Key), sizeof(Value)), sizeof(int));

In file included from /opt/rocm-5.0.0/include/thrust/sort.h:1358:
In file included from /opt/rocm-5.0.0/include/thrust/detail/sort.inl:26:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/sort.h:152:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/sort.inl:26:
In file included from /opt/rocm-5.0.0/include/thrust/find.h:381:
In file included from /opt/rocm-5.0.0/include/thrust/detail/find.inl:25:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/find.h:61:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/find.inl:19:
In file included from /opt/rocm-5.0.0/include/thrust/reduce.h:781:
In file included from /opt/rocm-5.0.0/include/thrust/detail/reduce.inl:28:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/reduce_by_key.h:87:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/generic/reduce_by_key.inl:37:
In file included from /opt/rocm-5.0.0/include/thrust/scan.h:1594:
In file included from /opt/rocm-5.0.0/include/thrust/detail/scan.inl:28:
In file included from /opt/rocm-5.0.0/include/thrust/system/detail/adl/scan.h:44:
In file included from /opt/rocm-5.0.0/include/thrust/system/hip/detail/scan.h:48:
In file included from /opt/rocm-5.0.0/include/rocprim/rocprim.hpp:33:
In file included from /opt/rocm-5.0.0/include/rocprim/intrinsics.hpp:29:
In file included from /opt/rocm-5.0.0/include/rocprim/intrinsics/thread.hpp:27:
/opt/rocm-5.0.0/include/rocprim/intrinsics/../detail/various.hpp:60:5: error: no template named 'enable_if_t' in na                                                                                          mespace 'std'; did you mean 'Enable_if_t'?
    std::enable_if_t<::rocprim::is_integral<T>::value && ::rocprim::is_integral<U>::value, int> = 0>
    ^~~~~
/opt/rocm-5.0.0/hip/include/hip/amd_detail/amd_hip_fp16.h:77:9: note: 'Enable_if_t' declared here
        using Enable_if_t = typename std::enable_if<cond, T>::type;

**Comment by Maetveis** [comment]
> @mfep I just tried building rocThrust against this PR and I get compilation errors. Can you confirm on your end?

@mfep Is on holiday, let me take this one. This should be fixed in rocThrust by: https://github.com/ROCmSoftwarePlatform/rocThrust/pull/220. Seems like we forgot upstreaming some small changes (mainly CI changes)

**Comment by stanleytsang-amd** [comment]
> > @mfep I just tried building rocThrust against this PR and I get compilation errors. Can you confirm on your end?
> 
> @mfep Is on holiday, let me take this one. This should be fixed in rocThrust by: [ROCmSoftwarePlatform/rocThrust#220](https://github.com/ROCmSoftwarePlatform/rocThrust/pull/220). Seems like we forgot upstreaming some small changes (mainly CI)

Thanks.  Verified rocThrust builds fine with the PR.

**Review by stanleytsang-amd — APPROVED** [review]

**Comment by stanleytsang-amd** [comment]
All tests pass now on CI.  Merging in.

**Pull request merged**